tests/functional/splinter_test.c

```diff
@@ -2383,7 +2383,9 @@ usage(const char *argv0)
       "\t%s --cache-per-table\n"
       "\t%s --parallel-perf --max-async-inflight [num] --num-pthreads [num] "
       "--lookup-positive-percent [num] --seed [num]\n"
-      "\t%s --num-bg-threads (number of background threads)\n"
+      "\t%s --num-normal-bg-threads (number of normal background threads)\n"
```

Comment:
I believe this was an old `splinter_test`-specific parameter that is now superceded by your better and more general handling in `config.[hc]`.  So I think this should go away entirely.

Reply:
I'm not sure I understand. `num-bg-threads` is being renamed to `num-normal-bg-threads` and the spelling error in the 2nd one's name is being fixed.

Not sure how all this can go away ... We should talk this thru.

Reply:
Look at your changes to config.c.  You added support for these command-line parameters to that file.

But this file used to have its own custom command-line parameter for controling bg threads.

Now that you've added support for bg-thread command-line parameters to config.c, you should remove the old special-case code from splinter_test.c and use the general code in config.c instead.

Reply:
@rtjohnso - I thought about your suggestion a bit more, and here is what I _think_ you are suggesting to do:

- Replace the inline code in `splinter_test()` with a call to `config_set_defaults(&master_cfg)`, followed by parsing cmd-line args into `master_cfg` using `config_parse()`.
- Then, pull out the bg-threads related values parsed into `master_cfg`, implementing something like `init_bg_threads_config_from_master_config()` [ like you've done elsewhere with `init_btree_config_from_master_config()` ]

Right? Well ... this won't quite work because after parsing these bg-thread args, `splinter_test()` does:

```
2646    if (splinter_test_parse_perf_args(&config_argv,
2647                                      &config_argc,
2648                                      &max_async_inflight,
...
```

I.e., we want to further parse more cmd-line arguments.

There are two problems with calling `config_parse()`:

1. It will eat up the bg-threads related config params, but will not advance argc / argv in the caller's context. We need a way to have `config_parse` advance `&config_argv, &config_argc` so that the subsequent call to `splinter_test_parse_perf_args()` is positioned correctly at the right args.
2. Calling `config_parse()` will mean caller can invoke with any other config params unrelated to bg-threads configuration. Those will be un-handled by this test, and will be misleading to the user. ( I see that this can be mitigated somewhat by a writing a new fn `init_bg_thread_config_from_master_config()`, but that is somewhat of a sub-optimal solution. )

For these reasons I did not implement your suggestion.

----
However, upon re-examination your comment did point out a flaw in my delta-commits.

In previous revs I did have one of the sub-cases in `unit/task_system_test.c` calling `config_parse()` to verify that the cmd-line arg parsing for bg-threads is working correctly.

As requested by @rosenhouse, and you agreed to that too, I've withdrawn that sequence and have inlined new test cases. That "lost" the code coverage for the parsing and error checking of bg-threads params that was added to `config _parse()`.

As it seems reasonable that we should go ahead with having bg-threads related in `master_config{}`, I'd like to retain the changes to `config_parse()`.  To verify that the right parsing is going on, as all this is not tested elsewhere, I added these small unit-test cases to `unit/config_parse_test.c`:

1. test_bg_threads_parse_configs_valid
2. test_bg_threads_parse_configs_invalid

These sub-cases do not require extra cmd-line args when the test is invoked and will provide required test coverage.

(Btw, general response is that all this config parsing stuff has to be overhauled separately through table driven logic. But that's not something I can take on right now.)

Reply:
Here is the problem with the arg parsing as it currently stands.  I'm pretty sure this is a real bug in the way you've changed things.

In the old code, `splinter_test.c` parsed its own command-line parameters, including --num-bg-threads, and then it called `test_config_parse` to parse the remaining parameters that are handled by `config.c`.

You have updated the `splinter_test.c` custom argument parsing to support two command line parameters, --num-normal-bg-threads and --num-memtable-bg-threads.

But you have also updated `config.c` to support the exact same command line parameters.

So now there are two difference pieces of code that could handle those command line parameters when someone runs splinter_test.  If the user puts --num-memtable-bg-threads or --num-normal-bg-threads early in their list of command line args, those arguments will get handled on lines 2617-2644 of splinter_test.c.  If they put those command line parameters later in the command-line arg list, those parameters will get processed by config.c

Now, when splinter_test.c initializes the task system on line 2795, it uses the locally declared array `num_bg_threads` (declared on line 2509).  This variable is set only by the command-line parsing code in splinter_test.c.  It is not configured by the command-line parsing code in config.c.

Thus, depending on where the user puts these command-line options, they could get silently ignored by splinter_test.c.

Hence my requested code change: delete the custom parsing code for those command-line args in splinter_test.c and convert the task_system initialization in splinter_test.c to use the values obtained from the config.c command-line-parsing code.

Reply:
Hmmm ... let me look at it some more later tomorrow. May be I did miss some double-parsing issue.

Reply:
@rtjohnso - I have reworked the code under [this recently uploaded commit](https://github.com/vmware/splinterdb/pull/469/commits/6e2b03cf43e330bebf383f4422f5e08e264c3e9f) to address this buggy behaviour.